kolibri/content/utils/annotation.py

```diff
@@ -1,34 +1,151 @@
-from kolibri.utils.time import local_now
+import logging as logger
+import math
+import os
+
 from django.conf import settings
+from kolibri.content.apps import KolibriContentConfig
+from kolibri.content.models import ChannelMetadata, ContentNode, File, LocalFile
+from sqlalchemy import and_, select
 
-from ..content_db_router import using_content_database
-from ..models import ChannelMetadata, ChannelMetadataCache
 from .channels import get_channel_ids_for_content_database_dir
+from .paths import get_content_file_name, get_content_storage_file_path
+from .sqlalchemybridge import Bridge
+
+logging = logger.getLogger(__name__)
 
+CONTENT_APP_NAME = KolibriContentConfig.label
 
 def update_channel_metadata_cache():
     """
-    After a channel is imported, or when the devserver is started,
+    If we are potentially moving from a version of Kolibri that did not import its content data,
     scan through the settings.CONTENT_DATABASE_DIR folder for all channel content databases,
     and pull the data from each database's ChannelMetadata object to update the ChannelMetadataCache
     object in the default database to ensure they are in sync.
     """
-    db_names = get_channel_ids_for_content_database_dir(settings.CONTENT_DATABASE_DIR)
-    # Delete ChannelMetadataCache objects in default db that are not found in CONTENT_DATABASE_DIR
-    ChannelMetadataCache.objects.exclude(id__in=db_names).delete()
-    # sync the ChannelMetadataCache objects in default db with ChannelMetadata objects in CONTENT_DATABASE_DIR
-    for db_name in db_names:
-        with using_content_database(db_name):
-            update_values = ChannelMetadata.objects.values()[0]
-
-        ch_metadata_obj, _ = ChannelMetadataCache.objects.update_or_create(
-            id=db_name,
-            defaults=update_values,
+    from .channel_import import import_channel_from_local_db
+    channel_ids = get_channel_ids_for_content_database_dir(settings.CONTENT_DATABASE_DIR)
+    for channel_id in channel_ids:
+        if not ChannelMetadata.objects.filter(id=channel_id).exists():
+            import_channel_from_local_db(channel_id)
+
+def set_leaf_node_availability_from_local_file_availability():
+    bridge = Bridge(app_name=CONTENT_APP_NAME)
```

Comment:
@ralphiee22 from here onwards in this file I am using the more SQL-y features of SQL Alchemy.